tech_reports/FlashAttention/FlashDecode.md

```diff
@@ -0,0 +1,186 @@
+
+**Non-causal mode**
+In non-causal mode, the query does not necessarily attend to all previous tokens. Users can pass in an arbitrary attention mask up to `kv_len`. Attention is computed on the entire length of KV cache, and the mask is applied to the attention scores. A use case for non-causal attention is in vision-language models, where cross attention is applied in the decode phase on the entire input image with arbitrary masking. An example of non-causal mode flash decode can be found in the [Llama 3 cross attention module](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tt/multimodal/llama_cross_attention.py).
+
+## 4 Performance Analysis
+
+### 4.1 Generic Performance
+
+We followed the same practice as in FlashAttention and measured TFLOPs/s for FlashDecode, with slight modification that we always compute the full attention up to the `kv_len` length, and the `seqlen` for the query is always 1. Therefore, the following FLOP formula is used:
+```
+FLOPs/s = 4 · bsz · kv_len · head_dim · n_q_heads / kernel_runtime
+```
+In general, FlashDecode is an memory bound operation. So in addition to FLOPs/s we also report the memory bandwidth utilization in GB/s. As a coarse estimate, we only consider the size of the KV cache, therefore the actually memory bandwidth utilization should be even higher than the number reported. The following formula is used to calculate the memory bandwidth utilization (in bytes per second, assuming BFP8 kv cache):
+```
+Bytes/s = 2 · bsz · head_dim · n_kv_heads · kv_len / kernel_runtime
+```
+
+We compare our implementation up to 16K sequence length with the baseline softmax attention implementation on TTNN, with input/output data in DRAM. Beyond 16K sequence length, the baseline implementation is not feasible due to the memory bandwidth limitation. We set up the experiment as follows:
+- BFLOAT16 query and BFP8 KV cache
+- batch size = 8, 16, 32
+- sequence length = 128k
+- head_dim = 64, 128, 256
+- n_kv_heads = 1
+- n_q_heads = 8
+
+![FlashDecodePerformanceTLOPs](images/FlashDecode/flashdecode_tflops.png)
+
+*Figure 3: FlashDecode performance in Compute (TFLOPs/s) as head dim, batch size, and kv cache length varies.*
+
+![FlashDecodePerformanceMemory](images/FlashDecode/flashdecode_gbs.png)
+
+*Figure 4: FlashDecode performance in Memory (GB/s) as head dim, batch size, and kv cache length varies.*
+
+The FlashDecode op is on average 2-10x faster than the baseline implementation. Compared to FlashAttention, FlashDecode is has lower speedup in terms of TFLOPs/s. This is expected because FlashDecode is an memory bound operation. On the other hand, FlashDecode achieved up to 180 Gb/s memory bandwidth utilization. The maximum observed memory bandwidth on Wormhole architecture is 250 Gb/s, so FlashDecode is hitting ~70% of that. Aside from memory and compute, another bottleneck is the noc traffic. We will discuss this in the next section.
```

Comment:
What would it take to use 250 GB/s? Maybe we need to look at how dram sharded matmuls are placed in core grid?

Reply:
I would suspect it to be the noc traffic. Maybe if ring reduction is used then it might be helpful!